User profile - release notes and more migration instructions (#27003)
keycloak/keycloak: docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc

= User Profile Changes

== User profile enabled by default

> User profile feature is now enabled by default. The feature `declarative-user-profile` is now not available anymore as the user profile is always considered enabled now.

Reviewer: ```suggestion
The user profile feature is now enabled by default. The  `declarative-user-profile`feature is no longer available, because the user profile is assumed to be always enabled.
```

> This means that realm switch `User Profile Enabled` is also not available and it was removed from the realm settings. Instead of `User Profile Enabled` switch, there is new switch `Unmanaged attributes` added to the realm settings.

Reviewer: ```suggestion
Therefore,  the *User Profile Enabled* switch is removed from the *Realm Settings* and replaced by *Unmanaged attributes*.
```

> When migrating from previous version, the behaviour is like this:

Reviewer: ```suggestion
When migrating from previous version, the behavior is as follows:
```

> * For the deployments where the realm switch `User Profile Enabled` was set to ON, the `Unmanaged attributes` switch will be set to `Disabled` after the migration. This means that just user attributes supported explicitly

Reviewer: ```suggestion
* For the deployments where *User Profile Enabled* was set to ON, the *Unmanaged attributes* will be set to *Disabled* after the upgrade. As a result, only user attributes supported explicitly
```

by user profile will be allowed.

> * For the deployments where the realm switch `User Profile Enabled` was set to OFF (which was also the default settings for the deployments with `declarative-user-profile` feature disabled, which was the default),

Reviewer: ```suggestion
* For the deployments where the *User Profile Enabled* was set to *OFF* (which was also the default settings for the deployments with `declarative-user-profile` feature disabled, which was the default),
```

> the `Unmanaged attributes` switch will be `Enabled` after the migration. This means the behaviour should be pretty much the same as in previous versions with user profile disabled. There will be still

Reviewer: ```suggestion
*Unmanaged attributes* will be set to *ON*` after the upgrade. As a result, the behavior should be basically the same as in previous versions with the user profile disabled. 
```

> `Attributes` tab available in user details in the admin console. Also users themselves are able to set arbitrary attributes through UI pages like registration page, update profile page etc as long as

Reviewer: ```suggestion
The *Attributes* tab will remain in user details in the Admin Console. Also users can now set arbitrary attributes through UI pages such as the registration page and update profile page  as long as 
```

> particular custom themes support it. The custom themes should work like before as well. However it is recommended to possibly update your themes to leverage user-profile (See below in the themes section) and

Reviewer: ```suggestion
a particular custom theme supports it. The custom themes should work as before as well. However, consider updating your themes to use the user-profile (See below in the themes section) and
```

> maybe even remove your custom themes at all if the only reason for them was adding custom attributes. It is also recommended to switch `Unmanaged attributes` to `Disabled` or enabled just for

Reviewer: ```suggestion
maybe even remove your custom themes at all if the only reason for them was adding custom attributes. We also recommend switching `Unmanaged attributes` to `Disabled` or enabled just for
```

> administrators if possible and rather rely on managed attributes as much as possible.

Reviewer: ```suggestion
administrators so that you can rely on managed attributes as much as possible.
```

== Breaking changes to the User Profile SPI

If you are using the User Profile SPI in your extension, you might be impacted by the API changes introduced in this release.

The `org.keycloak.userprofile.Attributes` interface includes the following changes:

* Method `getValues` was renamed to `get` to make it more aligned with the same operation from a regular Java `Map`
* Method `isRootAttribute` was moved to the utility class `org.keycloak.userprofile.UserProfileUtil.isRootAttribute`
* Method `getFirstValue` was renamed to `getFirst` to make it less verbose
* Method `getReadable(boolean)` was removed and now all attributes (including root attributes) are returned whenever they have read rights.

== Changes to Freemarker templates to allow rendering pages based on the user profile configuration set to a realm

In this release, the following templates were updated to make it possible to dynamically render attributes based
on the user profile configuration set to a realm:

* `login-update-profile.ftl`
* `register.ftl`
* `update-email.ftl`

> These templates are responsible for rendering the update profile (when the `Update Profile` required action is enabled to a user),

Reviewer: ```suggestion
These templates are responsible for rendering the update profile (when the *Update Profile* required action is enabled to a user),
```

> the registration, and the update email (when the `UPDATE_EMAIL` feature is enabled) pages, respectively.

Reviewer: ```suggestion
the registration, and the update email (when the *UPDATE_EMAIL* feature is enabled) pages, respectively.
```

> If you use a custom theme to change these templates, they will function as expect because only the content is updated.

Reviewer: ```suggestion
If you use a custom theme to change these templates, they will function as expected because only the content is updated.
```

However, we recommend you to take a look at how to configure a link:{adminguide_link}#user-profile[{declarative user profile}] and possibly avoid
changing the built-in templates by using all the capabilities provided by this feature.

Also, the templates used by the `declarative-user-profile` feature to render the pages for the same flows are longer necessary and removed in this release:

* `update-user-profile.ftl`
* `register-user-profile.ftl`

> If you were using the `declarative-user-profile` feature in previously releases with customizations to the above templates,

Reviewer: ```suggestion
If you were using the `declarative-user-profile` feature in a previous release with customizations to the above templates,
```

update the `login-update-profile.ftl` and `register.ftl` accordingly.

> == The update profile page when logging in for the first time through a broker now have its own Freemarker templates

Reviewer: ```suggestion
== The update profile page when logging in for the first time through a broker now has its own Freemarker templates
```

In this release, the server will render the update profile page when the user is authenticating through a broker for the
first time using the `idp-review-user-profile.ftl` template.

In previous releases, the template used to update the profile during the first broker login flow was the `login-update-profile.ftl`, the same used

> to update profile when users are authenticating to a realm.

Reviewer: ```suggestion
to update the profile when users are authenticating to a realm.
```
